File: javastory.budgetsh.stage1/src/javastory/budgetsh/stage1/step4/ui/console/CashBookConsole.java
package javastory.budgetsh.stage1.step4.ui.console;

import javastory.budgetsh.stage1.step1.entity.budget.CashBook;
import javastory.budgetsh.stage1.step1.entity.travel.Travel;
import javastory.budgetsh.stage1.step1.share.DatePair;
import javastory.budgetsh.stage1.step1.share.IdName;
import javastory.budgetsh.stage1.step1.share.Socialian;
import javastory.budgetsh.stage1.step4.logic.ServiceLogicLycler;
import javastory.budgetsh.stage1.step4.service.CashBookService;
import javastory.budgetsh.stage1.util.ConsoleUtil;
import javastory.budgetsh.stage1.util.Narrator;
import javastory.budgetsh.stage1.util.TalkingAt;

public class CashBookConsole {
	//
	private ConsoleUtil consoleUtil;
	private Narrator narrator;

	private CashBookService cashbookService;

	public CashBookConsole() {
		//
		narrator = new Narrator(this, TalkingAt.Left);
		consoleUtil = new ConsoleUtil(narrator);

		cashbookService = ServiceLogicLycler.shareInstance().createCashBookService();
	}

	public CashBook register() {
		//
		CashBook cashbook = null;
		Travel travel;
		Socialian socialian;

		while (true) {
			//
			String bankAccount = consoleUtil.getValueOf("\n Cashbook Bank Account(0.cashbook menu)");
			if (bankAccount.equals("0") || bankAccount.equals("") || bankAccount == null) {
				break;
			}
			if (cashbookService.exist(bankAccount)) {
				narrator.sayln("already exist this Account. -->" + bankAccount);
				continue;
			}

			String travelName = consoleUtil.getValueOf("\n Travel Name");
			String clubName = consoleUtil.getValueOf("\n club Name");
			travel = new Travel(travelName);
			travel.setClub(new IdName(travel.getId(), clubName));

			String leaderSocialId = consoleUtil.getValueOf("\n SocialId");
			String leaderFirstName = consoleUtil.getValueOf("\n First Name");
			String leaderFamilyName = consoleUtil.getValueOf("\n FamilyName");
			String leaderEmail = consoleUtil.getValueOf("\n Leader Email");
			String leaderPhone = consoleUtil.getValueOf("\n Leader Phone Number");
			socialian = new Socialian(leaderSocialId, leaderFirstName, leaderFamilyName, leaderEmail);
			socialian.setPhone(leaderPhone);

			double monthlyDue = Double.parseDouble(consoleUtil.getValueOf("\n monthlyDue"));
			double budgetGoal = Double.parseDouble(consoleUtil.getValueOf("\n Budget Goal"));
			String startDate = consoleUtil.getValueOf("\n startDate(ex >2017-03-23)");
			String endDate = consoleUtil.getValueOf("\n End Date(ex > 2017-04-04)");

			cashbook = new CashBook(travel, monthlyDue, budgetGoal, new DatePair(startDate, endDate));
			cashbook.setBankAccount(bankAccount);
			cashbook.setClub(travel.getClub());
			cashbook.setLeader(socialian);

			cashbookService.regist(cashbook);
			narrator.sayln("Regist -->" + cashbook);
			break;
		}
		return cashbook;

	}

	public CashBook find() {
		//
		CashBook cashbook = null;

		String account = consoleUtil.getValueOf("\n Cashbook Account(0.cashbook Menu)");
		if (account.equals("0") || account.equals("") || account == null) {
			return null;
		}

		if (cashbookService.exist(account)) {
			cashbook = cashbookService.retrieve(account);
			narrator.sayln("\n > Found Cashbook" + cashbook);
			return cashbook;
		} else {
			narrator.sayln(" > No cashbook in storage -->" + account);
		}

		return cashbook;
	}

	public CashBook modify() {
		//
		CashBook foundCashbook = find();
		String temp;

		double monthlyDue;
		double budgetGoal;
		String endDate;

		if (foundCashbook != null) {
			if (!(temp = consoleUtil.getValueOf("\n >monthly Due (Enter. no change):")).equals("")) {
				monthlyDue = Double.parseDouble(temp);
			} else {
				monthlyDue = foundCashbook.getMonthlyDue();
			}
			if (!(temp = consoleUtil.getValueOf("\n >budgetGoal (Enter. no change):")).equals("")) {
				budgetGoal = Double.parseDouble(temp);
			} else {
				budgetGoal = foundCashbook.getBudgetGoal();
			}

			endDate = consoleUtil.getValueOf("\n > End Date (Enter. no change): ");
			if (endDate.equals("")) {
				endDate = foundCashbook.getTerm().getEndDate();
			}

			foundCashbook.setTerm(new DatePair(foundCashbook.getTerm().getStartDate(), endDate));
			foundCashbook.setBudgetGoal(budgetGoal);
			foundCashbook.setMonthlyDue(monthlyDue);

			cashbookService.update(foundCashbook);

		} else {
			narrator.sayln("\n >No cashbook in the storage");
		}

		return foundCashbook;
	}

	public void remove() {
		//
		CashBook foundCashbook;

		String bankAccount = consoleUtil.getValueOf("\n Cashbook Bank Account(0.cashbook menu)");
		if (bankAccount.equals("0") || bankAccount.equals("") || bankAccount == null) {
			return;
		}
		if (cashbookService.exist(bankAccount)) {
			foundCashbook = cashbookService.retrieve(bankAccount);
			cashbookService.remove(foundCashbook);
			narrator.sayln("removed -->" + foundCashbook.getBankAccount());
		}
	}

}

File: javastory.budgetsh.stage5.1/src/step1/ui/console/CashBookConsole.js
const CashbookService = require('../../logic/CashBookService.js');
const Travel = require('../../domain/travel/Travel.js');
const IdName = require('../../domain/share/IdName.js');
const Socialian = require('../../domain/share/Socialian.js');
const CashBook = require('../../domain/budget/CashBook.js');
const DatePair = require('../../domain/share/DatePair.js');
const MainMenu = require('../menu/MainMenu.js');
var co = require('co');
var prompt = require('co-prompt');

exports.register = function(previousMenu) {
	//
	let bankAccount = '';
	let travelName = '';
	let clubName = '';
	let travel = null;
	let leaderSocialId = '';
	let leaderFirstName = '';
	let leaderFamilyName = '';
	let leaderEmail = '';
	let leaderPhone = '';
	let socialian = null;
	let monthlyDue = 0;
	let budgetGoal = 0;
	let startDate = '';
	let endDate = '';
	let cashbook = null;
	
	co(function*() {
		bankAccount = yield prompt('\n Cashbook Bank Account: ');
		if (CashbookService.exist(bankAccount)) {
			console.log("already exist this Account. -->" + bankAccount);
			MainMenu.selectCashbookMenu();
		}
		travelName = yield prompt('\n Travel Name: ');
		clubName = yield prompt('\n club Name: ');
		travel = new Travel(travelName);
		travel.club = new IdName(travel.id, clubName);

		leaderSocialId = yield prompt('\n SocialId: ');
		leaderFirstName = yield prompt('\n First Name: ');
		leaderFamilyName = yield prompt('\n FamilyName: ');
		leaderEmail = yield prompt('\n Leader Email: ');
		leaderPhone = yield prompt('\n Leader Phone: ');
		socialian = new Socialian(leaderSocialId, leaderFirstName, leaderFamilyName, leaderEmail);
		socialian.phone = leaderPhone;

		monthlyDue = Number(yield prompt('\n monthlyDue: '));
		budgetGoal = Number(yield prompt('\n Budget Goal: '));
		startDate = yield prompt('\n startDate(ex>2017-03-23)');
		endDate = yield prompt('\n End Date(ex>2018-01-10)');

		cashbook = new CashBook(travel, monthlyDue, budgetGoal, new DatePair(startDate, endDate));
		cashbook.bankAccount = bankAccount;
		cashbook.club = travel.club;
		cashbook.leader = socialian;

		CashbookService.regist(cashbook);
		
		console.log('Regist -->');
		cashbook.cashbook();
		MainMenu.selectCashbookMenu();
		return cashbook;
	});
}

exports.find = function() {

	//
	let bankAccount = '';
	let cashbook = null;
	
	co(function*() {
		bankAccount = yield prompt('\n Cashbook Account: ');
		cashbook = CashbookService.retrieve(bankAccount);
		if (cashbook == null) {
			console.log('> No Cashbook in storage -->' + bankAccount);
			MainMenu.selectCashbookMenu();
			return null;
		}
		console.log('Found Cashbook --> ');
		cashbook.cashbook();
		MainMenu.selectCashbookMenu();
		return cashbook;
	});
}

exports.modify = function() {
	let foundCashbook = null;
	let bankAccount = '';
	let monthlyDue = 0;
	let endDate = '';
	let budgetGoal = 0;
	let term = null;

	co(function*() {
			bankAccount = yield prompt('\n Cashbook Account: ');
			foundCashbook = CashbookService.retrieve(bankAccount);
			if(foundCashbook == null){
				console.log('> No Cashbook in storage.');
				MainMenu.selectCashbookMenu();
				return;
			}
			monthlyDue = parseInt(yield prompt('\n >monthly Due: '));
			budgetGoal = parseInt(yield prompt('\n >BudgetGoal: '));
			endDate = yield prompt('\n >End Date: ');

			term = new DatePair(foundCashbook.term.startDate, endDate);
			foundCashbook.term = term;
			foundCashbook.budgetGoal = budgetGoal;
			foundCashbook.monthlyDue = monthlyDue;

			if(CashbookService.update(foundCashbook)){
				console.log('Updated --> ');
				foundCashbook.cashbook();
			}else{
				console.log('fail to update.');
			}
			
			MainMenu.selectCashbookMenu();
	});
}

exports.remove = function(previousMenu) {
	var foundCashbook = null;
	let bankAccount = '';

	co(function*() {
		bankAccount = yield prompt('\n Cashbook BankAccount: ');
		if (CashbookService.exist(bankAccount)) {
			foundCashbook = CashbookService.retrieve(bankAccount);
			if(CashbookService.remove(foundCashbook)){
				console.log('removed --> ');
				foundCashbook.cashbook();
			}else{
				console.log('fail to remove');
			}

			MainMenu.selectCashbookMenu();
		}
	});
}
